library/alloc/src/rc.rs

```diff
@@ -2776,8 +2776,16 @@ impl<T: ?Sized> Weak<T> {
     ///
     /// # Safety
     ///
-    /// The pointer must have originated from the [`into_raw`] and must still own its potential
-    /// weak reference, and `ptr` must point to a block of memory allocated by the global allocator.
+    /// The pointer must have originated from the [`into_raw`] or [`into_raw_and_alloc`] functions
+    /// and must still own its potential weak reference, and `ptr` must point to a block of
+    /// memory allocated by `alloc`.
+    ///
+    /// Note that `from_raw` expects values that actually originated from a call to one of these
+    /// functions, not what these functions can maybe return, or are documented to potentially return.
+    /// For example, [`into_raw`] can return dangling pointers, but this doesn't allow you to create
+    /// a dangling pointer yourself and pass it to `from_raw`. This rule holds even if that pointer
```

Comment:
This is better, thanks.

I am still somewhat concerned though, do we have to write a whole paragraph explaining the meaning of "must have originated from" every single time we use that phrase? Will people read this, then see "must have originated from" elsewhere where we do *not* have a whole paragraph about this (IMO rather clear) notion, and conclude "they had a clarification there, but not here, hence they obviously mean something else by 'must have originated' because otherwise they would have added the same clarification"?

There s a real danger with having too many clarifications, which is that people think they describe a special case rather than clarifying a general principle.

Reply:
> do we have to write a whole paragraph explaining the meaning of "must have originated from" every single time we use that phrase?

The obvious solution to that problem is to link the phrase to a centralized location (e.g. the rust reference) giving the explanation of what it means.

Reply:
Yeah I guess that would be the alternative here.

With my suggestion above applied, I can live with this PR, though I am on the fence about whether it truly is an improvement -- I'm happy to leave that to the libs-api reviewer.

Reply:
> I am still somewhat concerned though, do we have to write a whole paragraph explaining the meaning of "must have originated from" every single time we use that phrase?

The main problem here is that `as_ptr` (and `into_inner` linking to it) says that it can potentially return null or invalid pointers. We say this to be future proof, not because we return null pointers, but this creates this potential misunderstanding that I'm trying to guard against.

With valid allocations, it makes sense that you require the pointer to come from a specific function, to prevent the [malloc + delete problem](https://stackoverflow.com/a/10854244). But with invalid ones it seems weird to do that.

One can very easily assume that this was written only keeping the valid pointers in mind, but not the invalid ones. Often even specifications leave something unspecified and then implementers interpret it differently.

Of course there is other misunderstandings too, and yeah eventually you run into the situation where you are creating more misunderstandings than fixing, but I think this PR is improving the situation.

> There s a real danger with having too many clarifications, which is that people think they describe a special case rather than clarifying a general principle.

Good point, that's what the word "Note" at the start tries to accomplish.

Reply:
I think the `Weak::new` reference is good, but we could pare down to mostly just that suggestion. If you're trying to second-guess the pointer origin, this is probably what you actually need.